### PR TITLE
feat(plugin): add object API routes and models

### PR DESCRIPTION
**Pull request by JacopoDaeli**
Adding object routes and models.
Follow other PRs related: test setup, controller logic add on etc

It only implements `GET /objects/:name` route. Other routes mockup only.
Will add implement other routes later with a separate PR.

### CODE REVIEW

**Review by jcrugzz — COMMENTED** [review]
Some initial questions to discuss, overall it looks good. Fastify seems to have fairly robust plugins. I also get the want to use raw dynamo to have more granular control for how we use it but we can more easily configure interesting features into our own wrapper library. Its a trade off but we should see about it.

File: lib/plugins/object.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
+const fp = require('fastify-plugin');
+
+const OBJECTS_TABLE = 'warehouse-objects';
+const OBJECT_VARIANTS_TABLE = 'warehouse-object-variants';
+const OBJECTS_HISTORY_TABLE = 'warehouse-objects-history';
+
+module.exports = fp(async function (fastify) {
```
[the remark this answers is not in the document]
Comment (JacopoDaeli) [reply lib/plugins/object.js]: @jcrugzz our new models are fairly easy actually. Currently, I use [DynamoDB.DocumentClient](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB/DocumentClient.html) which simplify the DynamoDB value types situation. `Dynastar` seems an overkilling.
Also do not forget that we need atomic transactions here and very fancy conditional inserts...  Not even sure `Dynastar` would support that.

File: lib/plugins/dynamo.js [hunk 2]
```diff
@@ -1,7 +1,14 @@
 const fp = require('fastify-plugin');
+const { DynamoDB } = require('aws-sdk');
 
 module.exports = fp(async function (fastify) {
-  // TODO(jdaeli): create an AWS.DynamoDB instance.
-  const dynamo = {};
+  const dynamo = new DynamoDB.DocumentClient(fastify.config.dynamo);
   fastify.decorate('dynamo', dynamo);
+}, {
+  fastify: '3.x',
```
[the remark this answers is not in the document]
Comment (JacopoDaeli) [reply lib/plugins/dynamo.js]: Yep :)

File: lib/plugins/object.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    };
+    return dynamo.query(params).promise();
+  });
+
+  fastify.decorate('putObjectVariant', function ({
+    name,
+    version,
+    env,
+    variant = '_default',
+    expiration = null,
```
[the remark this answers is not in the document]
Comment (JacopoDaeli) [reply lib/plugins/object.js]: This is an optional value used internally to return or not a result. It's not the DynamoDB record TTL.

[the remark this answers is not in the document]
Comment (JacopoDaeli) [reply lib/plugins/object.js]: In short, if an assets expires, we still want to keep track of it :)

File: lib/plugins/object.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    const createdAt = new Date().toUTCString();
+    const id = `${name}_${version}_${env}`;
+    const params = {
+      Item: { id, name, version, env, variant, expiration, data, createdAt },
+      TableName: OBJECT_VARIANTS_TABLE
+    };
+    return dynamo.put(params).promise();
+  });
+
+  fastify.decorate('putObjectHistory', function ({
```
[the remark this answers is not in the document]
Comment (JacopoDaeli) [reply lib/plugins/object.js]: It's explained in my initial PDR, but in short I want to use it to keep a full history of the object promotions/rollbacks per environment.

File: lib/plugins/object.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    const now = new Date();
+    if (headVersion && !headTimestamp) {
+      headTimestamp = now.getTime();
+    }
+    const lastModified = now.toUTCString();
+    const params = {
+      Key: { name, env, latestVersion, headVersion, headTimestamp, lastModified },
+      TableName: OBJECTS_TABLE
+    };
+    return new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (JacopoDaeli) [reply lib/plugins/object.js]: Yep, grazie mille. Updated this.

File: lib/plugins/object.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    headVersion = null,
+    headTimestamp = null
+  }) {
+    const now = new Date();
+    if (headVersion && !headTimestamp) {
+      headTimestamp = now.getTime();
+    }
+    const lastModified = now.toUTCString();
+    const params = {
+      Key: { name, env, latestVersion, headVersion, headTimestamp, lastModified },
```
[the remark this answers is not in the document]
Comment (JacopoDaeli) [reply lib/plugins/object.js]: Yep, that's what `objects-history` table is about.

File: lib/config.js [hunk 7]
```diff
@@ -1,9 +1,20 @@
...
-  /* eslint-enable no-process-env */
-});
+module.exports = fp(
+  async function (fastify) {
+    /* eslint-disable no-process-env */
+    fastify.decorate('config', {
+      port: process.env.PORT || 6666,
+      dynamo: {
+        accessKeyId: process.env.AWS_ACCESS_KEY_ID, // optional
+        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY, // optional
```
Comment (decompil3d) [inline lib/config.js:10]: The AWS SDK should automatically pull from these vars, so you don't need to manually pass in auth.

File: lib/config.js [hunk 8]
```diff
@@ -1,9 +1,20 @@
...
+        region: process.env.AWS_REGION || 'us-west-2'
```
Comment (decompil3d) [inline lib/config.js:11]: Same on this

File: lib/plugins/object.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    fastify.decorate('getObject', async function ({ name, env }) {
+      const params = {
+        Key: { name, env },
+        TableName: OBJECTS_TABLE
+      };
+      const { Item: item } = await dynamo.get(params).promise();
+      return item;
+    });
+
+    fastify.decorate('putObject', function ({
```
Comment (decompil3d) [inline lib/plugins/object.js:20]: Should this be an `async function`?

File: lib/plugins/object.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+          env,
+          latestVersion,
+          headVersion,
+          headTimestamp,
+          lastModified
+        },
+        TableName: OBJECTS_TABLE
+      };
+      const { Item: item } = await dynamo.get(params).promise();
+      return item;
```
Comment (decompil3d) [inline lib/plugins/object.js:44]: Should this be using some other method besides `get`?

File: lib/plugins/object.js [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        ExpressionAttributeValues: {
+          ':hkey': `${name}_${version}_${env}`
+        },
+        TableName: OBJECT_VARIANTS_TABLE
+      };
+      const { Items: items } = await dynamo.query(params).promise();
+      return items;
+    });
+
+    fastify.decorate('putObjectVariant', function ({
```
Comment (decompil3d) [inline lib/plugins/object.js:98]: Should this be an `async` function?

File: lib/plugins/object.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+      const createdAt = new Date().toUTCString();
+      const id = `${name}_${version}_${env}`;
+      const params = {
+        Item: { id, name, version, env, variant, expiration, data, createdAt },
+        TableName: OBJECT_VARIANTS_TABLE
+      };
+      return dynamo.put(params).promise();
+    });
+
+    fastify.decorate('putObjectHistory', function ({
```
Comment (decompil3d) [inline lib/plugins/object.js:115]: `async`?

**Review by decompil3d — CHANGES_REQUESTED** [review]
Good start. I'd love to see JSDoc added to these new methods so we can have solid typing from the start.

File: lib/config.js (hunk 7 above)
Comment (JacopoDaeli) [reply lib/config.js]: Reason why I did this is to have `region=us-west-2` as default, but ok. Let's remove this.

File: lib/plugins/object.js (hunk 9 above)
Comment (JacopoDaeli) [reply lib/plugins/object.js:20]: We do not `await` anything here. We return the promise directly, so no.

File: lib/plugins/object.js (hunk 10 above)
Comment (JacopoDaeli) [reply lib/plugins/object.js:44]: I do not think so, why?

We have two other methods indeed: one for queryAll and one for batchGet.

File: lib/plugins/object.js (hunk 11 above)
Comment (JacopoDaeli) [reply lib/plugins/object.js:98]: We do not `await` anything here. We return the promise directly, so no.

File: lib/plugins/object.js (hunk 12 above)
Comment (JacopoDaeli) [reply lib/plugins/object.js:115]: We do not `await` anything here. We return the promise directly, so no.

File: lib/plugins/object.js (hunk 10 above)
Comment (jkeslin-godaddy) [reply lib/plugins/object.js]: This is the `putObject` method -- so presumably you'd want to put here instead of getting, right?

Comment (JacopoDaeli) [reply lib/plugins/object.js]: Sorry, I misread your comment. Yes, should be `put` indeed. Fixed.

File: lib/plugins/object.js [hunk 13]
```diff
@@ -40,7 +40,7 @@ module.exports = fp(
         },
         TableName: OBJECTS_TABLE
       };
-      const { Item: item } = await dynamo.get(params).promise();
+      const { Item: item } = await dynamo.put(params).promise();
```
Comment (decompil3d) [inline lib/plugins/object.js]: Since you now `await` in this function, it should be `async`.

**Review by decompil3d — CHANGES_REQUESTED** [review]

File: lib/plugins/object.js (hunk 13 above)
Comment (JacopoDaeli) [reply lib/plugins/object.js]: Check again :)

**Review by jkeslin-godaddy — APPROVED** [review]

**Review by decompil3d — APPROVED** [review]

**Pull request merged**